include/ada/url.h

```diff
@@ -37,23 +45,45 @@ namespace ada {
      * A URL’s path is either an ASCII string or a list of zero or more ASCII strings, usually identifying a location.
      * TODO: Change this to list.
      */
-    char* path;
+    std::string_view path;
 
     /**
      * A URL’s query is either null or an ASCII string. It is initially null.
      */
-    std::optional<char*> query;
+    std::optional<std::string_view> query;
 
     /**
      * A URL’s fragment is either null or an ASCII string that can be used for further processing on the resource
      * the URL’s other components identify. It is initially null.
      */
-    std::optional<char*> fragment;
+    std::optional<std::string_view> fragment;
 
     /**
      * Used for returning the validity from the result of the URL parser.
      */
     bool is_valid = true;
+
+    /**
+     * A validation error indicates a mismatch between input and valid input.
+     * User agents, especially conformance checkers, are encouraged to report them somewhere.
+     */
+    bool has_validation_error = false;
+
+    /**
+     * A URL includes credentials if its username or password is not the empty string.
+     * @return
+     */
+    bool includes_credentials() const {
+      return username != nullptr || password != nullptr;
```

Comment:
This is not going to be correct if usename (e.g.) is a string view. I think you want something like !username.empty().

Reply:
You're right. Thanks.